src/Shared/CommunicationsUtilities.cs

```diff
@@ -330,13 +330,12 @@ internal static long GetHostHandshake(HandshakeOptions nodeType)
 
             Trace("MSBUILDNODEHANDSHAKESALT=\"{0}\", msbuildDirectory=\"{1}\", nodeType={2}, FileVersionHash={3}", salt, toolsDirectory, nodeType, FileVersionHash);
 
-            //FileVersionHash (32 bits) is shifted 8 bits to avoid session ID collision
-            //nodeType (4 bits) is shifted just after the FileVersionHash
-            //nodeHandshakeSalt (32 bits) is shifted just after hostContext
-            //the most significant byte (leftmost 8 bits) will get zero'd out to avoid connecting to older builds.
-            //| masked out | nodeHandshakeSalt | hostContext |              fileVersionHash             | SessionID
-            //  0000 0000     0000 0000 0000        0000        0000 0000 0000 0000 0000 0000 0000 0000   0000 0000
-            long baseHandshake = ((long)nodeHandshakeSalt << 44) | ((long)nodeType << 40) | ((long)FileVersionHash << 8);
+            // FileVersionHash (32 bits) is shifted 8 bits to avoid session ID collision
+            // HandshakeOptions (4 bits) is shifted just after the FileVersionHash
+            // nodeHandshakeSalt (32 bits) is shifted just after HandshakeOptions
+            //     nodeHandshakeSalt    | HandshakeOptions |             fileVersionHash           | SessionID
+            //  0000 0000 0000 0000 0000        0000        0000 0000 0000 0000 0000 0000 0000 0000  0000 0000
+            long baseHandshake = (nodeHandshakeSalt << 44) ^ ((short)nodeType << 40) ^ (FileVersionHash << 8);
```

Comment:
I'm afraid that the casts to `long` in the original code were important. The result of shifting an `int` left by N bits is still an `int` and all but the lowest 5 bits of N are discarded.

E.g.

```
(nodeHandshakeSalt << 44)
```

is the same as

```
(nodeHandshakeSalt << 12)
```

here. The fact that it's being assigned to a `long` doesn't make a difference.